src/resolution/graph.rs

```diff
@@ -858,7 +863,7 @@ impl<'a, TNpmRegistryApi: NpmRegistryApi>
     // Some packages may resolves to themselves as a dependency. If this occurs,
     // just ignore adding these as dependencies because this is likely a mistake
     // in the package.
-    if child_id != parent_id {
+    if child_nv != parent_path.nv {
```

Comment:
Bug fix is here (we should just compare the nvs and not the ids) and mitigation for past lockfiles is above.